lib/src/crawl_async.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) 2017, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+final _empty = new Future.value(null);
+
+/// Finds and returns every node in a graph who's nodes and edges are
+/// asynchronously resolved.
+///
+/// Cycles are allowed. If this is an undirected graph the [children] function
+/// may be symmetric. In this case the [roots] may be any node in each connected
+/// graph.
+///
+/// [V] is the type of values in the graph nodes. [K] must be a type suitable
+/// for using as a Map or Set key. [children] should return the next reachable
+/// nodes.
+///
+/// There are no ordering guarantees. This is useful for ensuring some work is
+/// performed at every node in an asynchronous graph, but does not give
+/// guarantees that the work is done in topological order.
+Stream<V> crawlAsync<K, V>(Iterable<K> roots, FutureOr<V> Function(K) readNode,
+    FutureOr<Iterable<K>> Function(K, V) children) {
+  final crawl = new _CrawlAsync(roots, readNode, children);
+  crawl.run();
+  return crawl.result.stream;
+}
+
+class _CrawlAsync<K, V> {
+  final result = new StreamController<V>();
+
+  final FutureOr<V> Function(K) readNode;
+  final FutureOr<Iterable<K>> Function(K, V) children;
+  final Iterable<K> roots;
+
+  final _seen = new Set<K>();
+
+  _CrawlAsync(this.roots, this.readNode, this.children);
+
+  /// Add all nodes in the graph to [result] and return a Future which fires
+  /// after all nodes have been seen.
+  Future run() async {
+    await Future.wait(roots.map(_visit));
+    await result.close();
+  }
+
+  /// Resolve the node at [key] and output it, then start crawling all of it's
+  /// children.
+  Future _crawlFrom(K key) async {
+    var value = await readNode(key);
+    result.add(value);
+    var next = await children(key, value) ?? const [];
+    await Future.wait(next.map(_visit));
+  }
+
+  /// Synchronously record that [key] is being handled then start work on the
+  /// node for [key].
+  ///
+  /// The returned Future will complete only after the work for [key] and all
+  /// transitively reachable nodes has either been finished, or will be finished
+  /// by some other Future in [_seen].
+  Future _visit(K key) {
+    if (_seen.contains(key)) return _empty;
+    _seen.add(key);
+    return _crawlFrom(key);
```

Comment:
It does feel arbitrary how we wait for the crawl to finish in one case, and don't in the other.

Instead we could make this not return a future at all (or even get rid of this function entirely?), and instead just `Future.wait` the entire `seen` list at the very end (and that would become the result)?

Reply:
I'm having a hard time figuring out how we can make that work.

We rely on the guarantee stated in the Doc comment: The Future won't complete until the work for all children is done, or guaranteed to be done within the Future returned in some previous call to `_visit`. Without this Future tracking that guarantee we seem to get stuck: either we could lose some children because we haven't yet added them to `_seen` before we `Future.wait` it, or we can't handle cycles because two Futures in `_seen` are waiting on eachother.

We could plausibly `_Future.wait(_seen)` if we add some other mechanism to handle cycles, but I'm not sure it's worth it.

Reply:
Hmm ya the cycles cause an issue, I am fine with it as is.